include/fmt/format.h

```diff
@@ -1871,22 +1909,77 @@ auto write_float(OutputIt out, const DecimalFP& fp,
   });
 }
 
+template <typename T, FMT_ENABLE_IF(std::is_floating_point<T>::value)>
+FMT_CONSTEXPR20 bool isinf(T value) {
+  if (is_constant_evaluated()) {
+#if defined(__cpp_lib_bit_cast) && defined(__cpp_if_constexpr)
+    if constexpr (is_fast_float<T>::value) {
+      using floaty =
+          std::conditional_t<std::is_same<T, long double>::value, double, T>;
+      using uint = typename dragonbox::float_info<floaty>::carrier_uint;
+      constexpr auto significand_bits =
+          dragonbox::float_info<floaty>::significand_bits;
+      auto bits = std::bit_cast<uint>(value);
+      return (bits & exponent_mask<floaty>()) &&
+             !(bits & ((uint(1) << significand_bits) - 1));
+    } else {
+      FMT_ASSERT(false, "floating point type is not supported");
+    }
+#endif
+  }
+  return std::isinf(value);
+}
+
+template <typename T, FMT_ENABLE_IF(std::is_floating_point<T>::value)>
+FMT_CONSTEXPR20 bool isfinite(T value) {
+  if (is_constant_evaluated()) {
+#if defined(__cpp_lib_bit_cast) && defined(__cpp_if_constexpr)
+    if constexpr (is_fast_float<T>::value) {
+      using floaty =
+          std::conditional_t<std::is_same<T, long double>::value, double, T>;
+      using uint = typename dragonbox::float_info<floaty>::carrier_uint;
+      auto bits = std::bit_cast<uint>(value);
+      return (bits & exponent_mask<floaty>()) != exponent_mask<floaty>();
+    } else {
+      FMT_ASSERT(false, "floating point type is not supported");
+    }
+#endif
+  }
+  return std::isfinite(value);
+}
+
+template <typename T, FMT_ENABLE_IF(std::is_floating_point<T>::value)>
+FMT_INLINE FMT_CONSTEXPR bool signbit(T value) {
+  if (is_constant_evaluated()) {
+#ifdef __cpp_if_constexpr
+    if constexpr (is_fast_float<T>::value) {
+      auto bits = detail::bit_cast<uint64_t>(static_cast<double>(value));
+      return (bits & (uint64_t(1) << (num_bits<uint64_t>() - 1))) != 0;
+    } else {
+      FMT_ASSERT(false, "floating point type is not supported");
+    }
```

Comment:
It looks like no because the code under `is_constant_evaluated()` check is still a part of function code. Without optimization, it's not even omitted, according to Compiler Explorer. So this constant-evaluated code has to be valid, and when it contains static assert, which is false, then the "runtime" version of this function becomes ill-formed.

Reply:
Sure but at the very least we could drop the else branch because non-IEEE is very uncommon and we would get a reasonable error anyway from `signbit` below.

Reply:
Done

Reply:
I see you dropped it from isinf and isfinite but not here.

Reply:
> I see you dropped it from isinf and isfinite but not here.

Only because you wrote:

> we would get a reasonable error anyway from `signbit` below.

and it's `signbit`.

Reply:
If this line is about the `std::signbit()` call below, then we won't get a reasonable error from there in some cases when FP numbers are non-IEEE and `std::signbit()` manages to work with them.

Reply:
> we won't get a reasonable error from there in some cases when FP numbers are non-IEEE and std::signbit() manages to work with them

That seems fine to me. We'll either get an error if `signbit` is not constexpr or it will return the correct result.

Reply:
Removed also here.